### PR TITLE
Set precision while writing H5 files.

### LINKED ISSUE #50: Reduce precision of tensors while H5 Writing.
**Is your feature request related to a problem? Please describe.**
It annoys me that H5 files are huge even though these are just model outputs. It would be useful to reduce the precision of the tensors when we dump them to disk. Additionally, it might be nice to have a compression. 

**Describe the solution you'd like**
Have an additional parameter in `H5FileImageWriter` which can be used to change the precision.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by AjeyPaiK**
This PR makes the following changes:

1. We can now reduce the precision of tensors while writing them to disc using `H5FileImageWriter`.
2. The utility functions used by callbacks are now moved to ahcore.utils under callbacks.
3. The model outputs are logits of type `float32` and hence dumping them into disc creates large h5 files. This is optimised by adding an optional normalization function (one of softmax, sigmoid) to scale the logits appropriately. The user can also set a precision of their choice (either float32, float16 or uint8).
4. Some minor changes has been made to pre transfroms to handle empty tiles better.
5. These changes fix #50 and #51 and #48.

### CODE REVIEW

**Review by moerlemans — CHANGES_REQUESTED** [review]
The approach to go from float32 to float16 makes sense and is a handy feature when dealing with large files. But the approach to go to uint8 is not ideal I think. Using the sigmoid does not take into account what the output of your model is and thus it will most likely make all your model outputs either zero or one (/0 or 255 in uint8). This would also make it impossible to distinguish between which class is the maximum. If the model output is outside the interval of [-6.2, 6.2] the value will be 0 or 255, as models are not  constrained this does not make sense to me.

I think it would be more appropriate to scale with a min and max value from the data and then save this value as well. Another option could be to softmax everything (as I think stitching can still be done/ should actually be done with the probabilities right?). 

You have also added a multiplier, is this necessary or is uint8 always a multiplier of 255 and the other always 1.0

**Comment by AjeyPaiK** [comment]
@moerlemans Thanks for the comment. I added `NormalizationTypes` to leave this up to the user. I agree that passing all outputs through sigmoid may not be the best way to go.

File: ahcore/callbacks/h5_callback.py [hunk 1]
```diff
@@ -10,15 +10,15 @@
...
+from ahcore.callbacks.utils import InferencePrecision, _get_h5_output_filename, NormalizationType
 from ahcore.utils.data import DataDescription, GridDescription
 from ahcore.utils.io import get_logger
 from ahcore.utils.types import GenericArray
 from ahcore.writers import H5FileImageWriter
 
 
 class WriteH5Callback(Callback):
-    def __init__(self, max_queue_size: int, max_concurrent_writers: int, dump_dir: Path):
+    def __init__(self, max_queue_size: int, max_concurrent_writers: int, dump_dir: Path, normalization_type: str):
```
Comment (moerlemans) [inline ahcore/callbacks/h5_callback.py:21]: In here there is now only an argument for the normalization and not for the precision. The precision is now set to uint8 without choice to the user. I think an extra parameter will be necessary, but then we add many arguments. (maybe also not so nice @jonasteuwen would you add more arguments here/another way to sovle this?)

Comment (AjeyPaiK) [reply ahcore/callbacks/h5_callback.py:21]: @jonasteuwen I like Marek's comment above. I think we should make the precision as a consumable parameter of the H5 callback's constructor.

Comment (AjeyPaiK) [reply ahcore/callbacks/h5_callback.py]: @moerlemans it's fixed now. Are there any other comments?

**Review by moerlemans — APPROVED** [review]
The PR adds multiple ways to make the datafootprint of h5 files lower. The changes are discussed in conversations. Standard values for the functions are such that there is no loss of data. The changes are also tested. LGTM

File: ahcore/writers.py [hunk 2]
```diff
@@ -86,6 +89,8 @@ def __init__(
     def init_writer(self, first_coordinates: GenericArray, first_batch: GenericArray, h5file: h5py.File) -> None:
         """Initializes the image_dataset based on the first tile."""
         if self._is_compressed_image:
+            if self._precision is not None:
```
Comment (EricMarcus-ai) [inline ahcore/writers.py:92]: Why is this check performed here? Can't it be done in the constructor already?

File: ahcore/writers.py [hunk 3]
```diff
@@ -170,6 +175,8 @@ def init_writer(self, first_coordinates: GenericArray, first_batch: GenericArray
             "format": _format,
             "dtype": str(first_batch.dtype),
             "is_binary": self._is_compressed_image,
+            "precision": self._precision.value if self._precision else "FP32",
```
Comment (EricMarcus-ai) [inline ahcore/writers.py:178]: Is this "FP32" correct? In the type class you have "float32", why not use your enum here? I.e. make it something like `str(InferencePresicion.FP32)`?

File: ahcore/writers.py [hunk 4]
```diff
@@ -170,6 +175,8 @@ def init_writer(self, first_coordinates: GenericArray, first_batch: GenericArray
...
+            "multiplier": self._precision.get_multiplier() if self._precision else 1.0,
```
Comment (EricMarcus-ai) [inline ahcore/writers.py:179]: Similar here, probably best to use the values from the actual Type you created, such that any changes in the future will be captured by changing the the type, instead of having to scout out these values as well.

File: ahcore/writers.py [hunk 5]
```diff
@@ -205,6 +212,12 @@ def consume(
         try:
             with h5py.File(self._filename.with_suffix(".h5.partial"), "w") as h5file:
                 first_coordinates, first_batch = next(batch_generator)
+
+                if self._precision:
```
Comment (EricMarcus-ai) [inline ahcore/writers.py:216]: Since you have the same code below, you could decide to place this into a small method like
```
def adjust_batch_precision(batch: Tensor) -> Tensor:
    "your code here"
    return batch
```
Then in both cases you can call `batch = self.adjust_batch_precision(batch)`.

File: config/callbacks/default.yaml [hunk 6]
```diff
@@ -32,6 +32,8 @@ model_summary:
 write_h5_callback:
   max_queue_size: 100
   max_concurrent_writers: 4
+  normalization_type: "sigmoid" # Make sure to use the appropriate normalization for your use case.
```
Comment (EricMarcus-ai) [inline config/callbacks/default.yaml:35]: Can you keep the defaults to no normalization and no compression?

File: ahcore/callbacks/tiff_callback.py [hunk 7]
```diff
@@ -200,7 +200,7 @@ def _write_tiff(
             mpp=h5_reader.mpp,
             tile_size=tile_size,
             pyramid=True,
-            compression=TiffCompression.CCITTFAX4,
+            compression=TiffCompression.ZSTD,
```
Comment (EricMarcus-ai) [inline ahcore/callbacks/tiff_callback.py:203]: What is the reason for this change?

File: ahcore/utils/callbacks.py [hunk 8]
```diff
@@ -27,6 +29,41 @@
...
+        else:
+            raise ValueError("Function not supported")
+
+
+class InferencePrecision(str, Enum):
+    FP16 = "float16"
+    FP32 = "float32"
+    UINT8 = "uint8"
+
+    def __str__(self) -> str:
```
Comment (EricMarcus-ai) [inline ahcore/utils/callbacks.py:53]: Is this necessary? When subclassing both to str and Enum, doesn't `str(InferencePrecision.FP16)` already return the string value?

**Review by EricMarcus-ai — COMMENTED** [review]
Nice additions, I have added some 'cosmetic' requests. I also very much appreciate the test that was added, only it seems it is not testing the new 'precision' functionality yet? Perhaps a small test of that kind could be added, to check if the precision/ normalization functionality works?

File: ahcore/writers.py (hunk 2 above)
Comment (AjeyPaiK) [reply ahcore/writers.py:92]: The subsequent lines of code (from line 93) are executed only if line 91 evaluates to True. So, we have to have the if condition of self._is_compressed_image here. 

Let's say we move it to the constructor of the class, it is possible, but we will have to check the if condition again at line 91.  So, we would be repeating `if self._is_compressed_image` two times. That's why, the precision is checked once while initializing the writer.

File: ahcore/writers.py (hunk 5 above)
Comment (AjeyPaiK) [reply ahcore/writers.py:216]: Ah that's a good one. How silly of me!

File: config/callbacks/default.yaml (hunk 6 above)
Comment (AjeyPaiK) [reply config/callbacks/default.yaml:35]: Right, yes indeed. That's best.

File: ahcore/callbacks/tiff_callback.py (hunk 7 above)
Comment (AjeyPaiK) [reply ahcore/callbacks/tiff_callback.py:203]: Both @moerlemans and I noticed this error and I raised an issue earlier [here](https://github.com/NKI-AI/ahcore/issues/48). When you set it, did you try it with multiprocessing?

File: ahcore/writers.py (hunk 2 above)
Comment (AjeyPaiK) [reply ahcore/writers.py:92]: Discussed this offline and what we have now seems okay.

**Review by EricMarcus-ai — APPROVED** [review]
